connector.go

```diff
@@ -0,0 +1,300 @@
+
+	channel, err := conn.Channel()
+	if err != nil {
+		return nil, fmt.Errorf(errMessage, err)
+	}
+
+	if prefetchCount > 0 {
+		err = channel.Qos(prefetchCount, 0, false)
+		if err != nil {
+			return nil, fmt.Errorf(errMessage, err)
+		}
+	}
+
+	return channel, nil
+}
+
+func connectionString(settings *ConnectionSettings) (string, error) {
+	const errMessage = "failed to create connection string: %w"
+
+	var builder strings.Builder
+
+	_, err := builder.WriteString("amqp://")
+	if err != nil {
+		return "", fmt.Errorf(errMessage, err)
+	}
+
+	if len(settings.UserName) > 0 || len(settings.Password) > 0 {
+		_, err = builder.WriteString(url.QueryEscape(settings.UserName))
+		if err != nil {
+			return "", fmt.Errorf(errMessage, err)
+		}
+
+		_, err = builder.WriteString(":")
+		if err != nil {
+			return "", fmt.Errorf(errMessage, err)
+		}
+
+		_, err = builder.WriteString(url.QueryEscape(settings.Password))
+		if err != nil {
+			return "", fmt.Errorf(errMessage, err)
+		}
+
+		_, err = builder.WriteString("@")
+		if err != nil {
+			return "", fmt.Errorf(errMessage, err)
+		}
+	}
+
+	_, err = builder.WriteString(url.QueryEscape(settings.Host))
+	if err != nil {
+		return "", fmt.Errorf(errMessage, err)
+	}
+
+	_, err = builder.WriteString(":")
+	if err != nil {
+		return "", fmt.Errorf(errMessage, err)
+	}
+
+	_, err = builder.WriteString(strconv.Itoa(settings.Port))
+	if err != nil {
+		return "", fmt.Errorf(errMessage, err)
+	}
+
+	_, err = builder.WriteString("/")
+	if err != nil {
+		return "", fmt.Errorf(errMessage, err)
+	}
+
+	return builder.String(), nil
+}
+
+func watchConnectionNotifications(conn *amqp.Connection, name string, initWG *sync.WaitGroup, closeWG *sync.WaitGroup) {
+	closeChan := conn.NotifyClose(make(chan *amqp.Error))
+	blockChan := conn.NotifyBlocked(make(chan amqp.Blocking))
+
+	initWG.Done()
+
+	for {
+		select {
+		case err := <-closeChan:
+			if err != nil {
+				//nolint: godox // follow-up task
+				// TODO: reconnect logic
+				slog.Debug("connection unexpectedly closed", "type", name, "cause", err)
+			}
+
+			if err == nil {
+				slog.Debug("closed connection", "type", name)
+
+				closeWG.Done()
+
+				return
+			}
+
+		case block := <-blockChan:
+			slog.Warn("connection exception", "type", name, "cause", block.Reason)
+		}
+	}
+}
+
+func watchChannelNotifications(channel *amqp.Channel, name string, returnHandler ReturnHandler, initWG *sync.WaitGroup, closeWG *sync.WaitGroup) {
+	closeChan := channel.NotifyClose(make(chan *amqp.Error))
+	cancelChan := channel.NotifyCancel(make(chan string))
+	returnChan := channel.NotifyReturn(make(chan amqp.Return))
+
+	initWG.Done()
+
+	for {
+		select {
+		case err := <-closeChan:
+			if err != nil {
+				//nolint: godox // follow-up task
+				// TODO: reconnect logic
+				slog.Debug("channel unexpectedly closed", "type", name, "cause", err)
+			}
+
+			if err == nil {
+				slog.Debug("closed channel", "type", name)
+
+				closeWG.Done()
+
+				return
+			}
```

Comment:
one of those error checks seems useless, if you swap the checks you can remove one check because of the early return